Masonry/View.jsx

```diff
@@ -120,44 +129,74 @@ export default class Masonry extends React.PureComponent {
     window.removeEventListener('resize', this.onResize);
   }
 
-  componentWillReceiveProps(nextProps) {
-    if (nextProps.items.length !== this.props.items.length) {
-      this.layout(nextProps);
+  componentDidUpdate (prevProps, prevState) {
+    if (prevProps.items.length !== this.props.items.length) {
+      this.layout(this.props, false)
     }
   }
 
   onResize = throttle(() => {
-    this.layout(this.props, true);
+    this.layout((this.state.props || this.props), true);
   }, 150, { trailing: true })
 
+  dynamicWidth = () => {
+    const _p = this.state.props || this.props
+    const _cols = _p.numColumns
+    const _containerWidth = this.node.offsetWidth
+    const _gutter = _p.columnGutter
+    if (this.state.containerWidth === _containerWidth && this.state._colWidth) {
+      return this.state._colWidth;
+    }
+    // Column width for will be what ever is left over after the total
+    // gutter has been subtracted from the container width, divided by
+    // number of columns.
+    // Number of gutters are one less than number of columns.
+    const _dynColWidth = Math.floor((_containerWidth - _gutter * (_cols - 1)) / _cols)
+    const _colWidth = _cols ? _dynColWidth : _p.columnWidth
+    if (!_colWidth) {
+      throw new Error(`Can't figure out column width, either 'numColumns' or 'columnWidth' needs to be set.`);
+    }
+    this.setState({
+      columnWidth: _colWidth,
+      containerWidth: _containerWidth,
+    })
+    return _colWidth
+  }
+
   layout(props, rearrange=false) {
     if (!this.node) {
       return;
     }
 
     const {
-      columnWidth,
+      numColumns,
       columnGutter,
       items,
       itemComponent
     } = props;
 
-    const componentName = itemComponent.constructor.displayName || itemComponent.constructor.name;
+    let c = itemComponent.constructor;
```

Comment:
Could this be abstracted to another function? I would like to see it remain constant 👍